src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/Diagnostics.cs

```diff
@@ -52,14 +54,36 @@ internal partial class RazorCustomMessageTarget
 
         var csharpDiagnostics = await csharpTask.ConfigureAwait(false);
         var htmlDiagnostics = await htmlTask.ConfigureAwait(false);
+        var csharpAdditionalDiagnostics = await csharpAdditionalTask.ConfigureAwait(false);
 
         if (csharpDiagnostics is null || htmlDiagnostics is null)
         {
             // If either is null we don't have a complete view and returning anything will cause us to "lock-in" incomplete info. So we return null and wait for a re-try.
             return null;
         }
 
-        return new RazorPullDiagnosticResponse(csharpDiagnostics, htmlDiagnostics);
+        return new RazorPullDiagnosticResponse(csharpDiagnostics, htmlDiagnostics, csharpAdditionalDiagnostics);
+    }
+
+    /// <summary>
+    /// Roslyn analyzers are allowed to report diagnostics for additional files, which can be .razor files in this case. We issue a request to Roslyn to see if they want to
+    /// report any diagnostics on the razor file. If so, those will be added to the csharp diagnostics being reported.
+    /// </summary>
+    private async Task<VSInternalDiagnosticReport[]?> GetCSharpAdditionalDiagnosticsAsync(VSTextDocumentIdentifier hostDocument, Guid correlationId, CancellationToken cancellationToken)
+    {
+        using var _ = _telemetryReporter.TrackLspRequest(VSInternalMethods.DocumentPullDiagnosticName, RazorLSPConstants.RazorCSharpLanguageServerName, correlationId);
+        var request = new VSInternalDocumentDiagnosticsParams
+        {
+            TextDocument = hostDocument,
+        };
+
+        var response = await _requestInvoker.ReinvokeRequestOnServerAsync<VSInternalDocumentDiagnosticsParams, VSInternalDiagnosticReport[]?>(
+            VSInternalMethods.DocumentPullDiagnosticName,
```

Comment:
I'm surprised this works, but I'm glad it does, and doesn't need a custom endpoint on the Roslyn side

An alternative I could imagine, since this code is on the "VS side" of things, would be to pull in the workspace via MEF and grab the solution, and the TextDocument, and ask for diagnostics directly.

Reply:
I _think_ it does, I'm still having trouble with getting the end to end verification set up. That's why this is a draft for now

Reply:
https://github.com/dotnet/roslyn/blob/d88c4d64fd7a026f14e1630ae8e7247ee77c21b4/src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs#L119 will make it so this doesn't work... either need to create a new endpoint, start sending didOpen/didClose events, or remove that restriction in roslyn. Going to ping the team and see what they think

Reply:
That is, sadly, what I expected. Also I would veto sending didOpen/didClose events. We'll get there eventually with cohosting, but doing it now seems either wrong, or a bug farm, depending on how its done.

Reply:
> An alternative I could imagine, since this code is on the "VS side" of things, would be to pull in the workspace via MEF and grab the solution, and the TextDocument, and ask for diagnostics directly.


The problem with this approach is that we'd have to rerun the analyzers ourselves, they're not stored as default diagnostics on the compilation we'd produce unless we also start tracking analyzer references etc. I think that's a pretty big pit to start falling into

Reply:
Agreed, I want nothing to do with the likes of that :)